### ISSUE TITLE
[Feature]: CodeBlock and Formula primitives for technical decks (syntax highlighting + LaTeX math)

### ISSUE THREAD

**Issue by Kevinweisl** [1weiho/open-slide#467]
### Problem

I teach programming and AI courses, and almost every deck I make needs two things open-slide has no primitive for today: a **code block with syntax highlighting** and a **math formula** (fractions, subscripts, sums, matrices).

**What exists today.** The public surface of `@open-slide/core` is `Page`, `Steps` / `Step`, `MorphElement`, `ImagePlaceholder`, `useSlidePageNumber`, `useIsActivePage`, the `design` tokens, `notes`, and `transition`. Nothing renders code or math, and the `slide-authoring` skill has no reference for either.

**What decks do instead.** I went through `apps/demo`:

- 16 of the 19 demo decks render code with a hand-rolled `<pre>` / `<code>` and an inline monospace font.
- 4 of them define their own code-rendering helpers. `nextjs-ppr-cache` is the clearest example: it declares `kw()`, `fn()`, `str()`, `com()`, `tag()` and colors every token by hand, one `<span>` at a time.
- `llm-fundamentals` (an AI deck) has to write formulas as plain monospace text, e.g. `p = softmax(logits)`. A fraction like `QK^T / sqrt(d_k)` is not expressible at all.

**Why agents can't fix this on their own.** The `slide-authoring` skill is explicit: *"Do not add dependencies. Only `react`, `@open-slide/core`, and standard web APIs are available."* So every agent-generated deck re-invents a tokenizer in `index.tsx`, the result varies from deck to deck, and math is effectively impossible. Since the framework's value is that the agent focuses on content, this feels like it belongs in core.

**Codebase check.** `packages/core` has no highlighter or math library (the `shiki` in the lockfile is pulled in by fumadocs for `apps/web` only). I also searched issues and discussions and found no prior request for code or math.

### Proposed solution

Two primitives exported from `@open-slide/core`, following the same API shape as `ImagePlaceholder` and `MorphElement`: a few semantic props, then `style`, `className`, and `...rest` spread onto the root element, with `style` merged last so every default can be overridden.

```tsx
import { CodeBlock, Formula, type Page } from '@open-slide/core';

const Attention: Page = () => (
  <div style={{ width: '100%', height: '100%', padding: 120, background: 'var(--osd-bg)' }}>
    <h2 style={{ fontSize: 80, margin: 0 }}>Scaled dot-product attention</h2>

    <Formula display style={{ fontSize: 56, margin: '48px 0' }}>
      {String.raw`\mathrm{Attention}(Q, K, V) = \mathrm{softmax}\left(\frac{QK^{\top}}{\sqrt{d_k}}\right) V`}
    </Formula>

    <CodeBlock lang="python" lineNumbers highlightLines={[2, 3]} style={{ fontSize: 30 }}>
      {`def attention(q, k, v):
    scores = q @ k.T / math.sqrt(k.shape[-1])
    weights = softmax(scores)
    return weights @ v`}
    </CodeBlock>
  </div>
);
```

**`CodeBlock`**

- Props: `lang`, `lineNumbers`, `highlightLines`, plus `style` / `className` / `...rest` (`HTMLAttributes<HTMLPreElement>`). Children is the source string. No `fontSize` or `theme` props: sizing and looks go through `style`, the way every other slide element works.
- Renders a `<pre>` with sensible defaults (28px monospace, dark surface, `--osd-radius` corners, padding), sized in absolute pixels, and it never scrolls. It follows the same 1080px vertical-budget rule as everything else.
- Token colors are read from `--osd-code-*` CSS variables (`keyword`, `string`, `comment`, `function`, `number`, `punctuation`) with built-in fallbacks. Set them on the page root or on the block's `style` to restyle, matching how `--osd-*` tokens already work. A light look is just a different `background` plus variables; no `variant` prop.
- Tokens are rendered as React `<span style>` elements, not injected HTML, so the block depends on no stylesheet or class names and stays inside the "inline styles only" rule for slides.
- `...rest` forwards `data-slide-loc`, so the inspector can target the block like it targets `ImagePlaceholder`.
- Composes with `<Steps>` / `<Step>` as-is for progressive reveal. A line-range `reveal` API could follow later; out of scope for the first PR.

**`Formula`**

- Props: `display` (block vs inline), plus `style` / `className` / `...rest`. Children is a LaTeX string. Color and size inherit from the wrapper, so `style={{ color, fontSize }}` is all a deck needs.
- Named `Formula` rather than `Math` on purpose: `import { Math } from '@open-slide/core'` would shadow the global `Math` object inside the slide module, and `Math.sqrt` in the same file would break.

**Keeping it light at runtime**

- One module-level highlighter instance, grammars loaded once per language, and results memoized by `(lang, code)`. The thumbnail rail, overview grid, and presenter render many pages at once, so per-instance highlighting would be wasteful.
- First paint shows the plain monospace text; token colors fill in without layout shift because the `<pre>` metrics are identical. Exports wait for the highlighted state (see below).
- The JS regex engine is used, so no WASM is shipped or fetched.

**Agent skills and docs**

- Add `references/code-block.md` and `references/formula.md` to the `slide-authoring` skill, a row for each in the primitive table, and matching self-review items. This is what makes the primitives reachable from `create-slide`.
- Add `docs/primitive/code-block.mdx` and `docs/primitive/formula.mdx` on the site.
- Add a `code-and-math` demo deck to `apps/demo` for dogfooding and screenshots.

**Dependency options** — core is careful about install size, so here is what I compared (npm `unpackedSize`, current versions):

| Code highlighting | Size | Trade-offs |
| --- | --- | --- |
| `@shikijs/core` + `@shikijs/engine-javascript` + per-language grammars from `@shikijs/langs` (recommended) | core 64 KB, engine 11 KB, each grammar loaded on demand via dynamic `import()` | VS Code-grade TextMate grammars; async, so the component signals readiness for exports (see below); no WASM since the JS regex engine is used |
| `sugar-high` | 95 KB, sync | Tiny and simple, but a single JS-flavoured grammar. Python, C, SQL, and shell decks would look wrong |
| No dependency: a small hand-written tokenizer for a handful of languages | 0 | Zero install cost, but low fidelity and a grammar to maintain per language |

| Math | Size | Trade-offs |
| --- | --- | --- |
| `katex` (recommended) | 4.0 MB unpacked, dominated by font files; only the woff2 set is fetched at runtime | Sync, mature, HTML + CSS output that prints and screenshots reliably; its stylesheet is scoped under `.katex`; fonts must ship with the static build |
| `temml` | 2.2 MB unpacked, ~150 KB runtime | Outputs MathML and relies on browser MathML Core plus a math font; needs verification through the `html-to-image` path used by PPTX export |
| `mathjax-full` | 34 MB | Too heavy for a runtime that ships to every user |

Both recommended libraries would be lazy-loaded with dynamic `import()`, following the existing precedent for `fflate` and `html-to-image` in `export-html.ts` / `export-pptx.ts`, so decks without code or math pay nothing.

**Runtime and export compatibility** — things I verified in the codebase that shape the design:

- PDF and PPTX export (`print-ready.ts`) already wait on `document.fonts.ready`, on `[data-waitfor]` selectors, and on animations settling. An async highlighter can set `data-waitfor` on its root, so exports capture the highlighted state rather than the raw text. KaTeX is sync and needs nothing extra.
- HTML export collects CSS from `document.styleSheets`, so the KaTeX stylesheet is carried along. KaTeX fonts need to be reachable from the static build, so they would be bundled as assets.
- Slides render in the same document as the viewer, so Tailwind's preflight already reaches slide content. KaTeX and Tailwind preflight coexist on many docs sites, but I would verify it in the demo deck rather than assume it.
- The inspector only tags host elements plus an allowlist of forwarding components (`loc-tags-plugin.ts`, currently just `ImagePlaceholder`). The new primitives would join that list so click-to-comment works on them, while inline text editing would be disabled inside since the source is a single string.
- `DesignFonts` only has `display` and `body`. Adding a `mono` token (`--osd-font-mono`) so the Design panel can tweak the code font seems natural, but could be its own PR.

### Alternatives considered

- **Keep hand-rolling per deck** (status quo). Works for JS-heavy decks written by hand, but every agent run re-invents it, and math stays out of reach.
- **Skill-only guidance.** Add a reference to `slide-authoring` on how to hand-roll a code block. Cheaper, but still no math, and the highlighting quality stays tied to whatever the agent improvises.
- **Allow slides to add dependencies.** Contradicts the skill's hard rule and the model where the agent never touches `package.json`.

### Which package would this affect?

- @open-slide/core (runtime, inspector, present mode, CLI)
- Agent skills (create-slide, slide-authoring, apply-comments)
- Docs / marketing site

### Additional context

I would like to implement this. Proposed split so each PR stays reviewable:

1. `CodeBlock` + skill references + docs + demo deck (`@open-slide/core` minor changeset).
2. `Formula` on top of the same plumbing (readiness signal, inspector allowlist, docs pattern).

Questions before I start:

1. Is a lazy-loaded highlighter dependency acceptable for core, or would you rather start with a zero-dependency tokenizer?
2. Any preference on naming (`CodeBlock` / `Formula`) or on adding `fonts.mono` to `DesignSystem`?
3. Is the two-PR split the shape you'd want, or one PR for both?

**Comment by Kevinweisl** [comment]
Opened #468 (draft) with the `CodeBlock` half of this proposal. The PR description has the API, the measured bundle impact (entry chunk unchanged; shiki and each grammar code-split and loaded on first use), and the inspector/export handling. The CI workflow passes on my fork for the same head: https://github.com/Kevinweisl/open-slide/actions/runs/34685931193

`Formula` would follow as a separate PR once the dependency direction in #468 is settled.
